gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/AsNumberStep.java

```diff
@@ -0,0 +1,192 @@
+    public int hashCode() {
+        int result = super.hashCode();
+        result = 31 * result + (numberToken != null ? numberToken.hashCode() : 0);
+        return result;
+    }
+
+    @Override
+    public AsNumberStep<S> clone() {
+        final AsNumberStep<S> clone = (AsNumberStep<S>) super.clone();
+        clone.numberToken = this.numberToken;
+        return clone;
+    }
+
+    @Override
+    public String toString() {
+        return StringFactory.stepString(this);
+    }
+
+    public static Number parseNumber(final String value) {
+        try {
+            boolean isFloatingPoint = value.contains(".") || value.contains("e") || value.contains("E");
+            if (isFloatingPoint) {
+                BigDecimal result = new BigDecimal(value.trim());
+                if (BigDecimal.valueOf(result.doubleValue()).compareTo(result) == 0) {
+                    return result.doubleValue();
+                }
+                return result;
+            }
+            BigInteger result = new BigInteger(value.trim());
+            if (result.bitLength() <= 31) { // default to int if not specified, smaller sizes need to be intentionally set
+                return result.intValue();
+            } else if (result.bitLength() <= 63) {
+                return result.longValue();
+            }
+            return result;
+        } catch (NumberFormatException nfe) {
+            throw new NumberFormatException(String.format("Can't parse string '%s' as number.", value));
+        }
+    }
+
+    private static Number autoNumber(Number number) {
+        final Class<? extends Number> clazz = number.getClass();
+        if (clazz.equals(Float.class)) {
+            return castNumber(number, N.ndouble);
+        }
+        return number;
+    }
+
+    private static Number castNumber(Number number, N numberToken) {
+        int sourceBits = getNumberBitsBasedOnValue(number);
+        int targetBits = getNumberTokenBits(numberToken);
+        if (sourceBits > targetBits) {
+            throw new ArithmeticException(String.format("Can't convert number of type %s to %s due to overflow.",
+                    number.getClass().getSimpleName(), numberToken.toString()));
+        }
+        if (numberToken == N.nbyte) {
+            return number.byteValue();
+        } else if (numberToken == N.nshort) {
+            return number.shortValue();
+        } else if (numberToken == N.nint) {
+            return number.intValue();
+        } else if (numberToken == N.nlong) {
+            return number.longValue();
+        } else if (numberToken == N.nfloat) {
+            return number.floatValue();
+        } else if (numberToken == N.ndouble) {
+            return number.doubleValue();
+        } else if (numberToken == N.nbigInt) {
+            return BigInteger.valueOf(number.longValue());
+        } else if (numberToken == N.nbigDecimal) {
+            return BigDecimal.valueOf(number.doubleValue());
+        }
+        return number;
```

Comment:
should consider throwing here, this could be a future footgun if another `N` enum is added and number is directly returned.